lib/connection.cpp

```diff
@@ -1140,6 +1140,22 @@ JoinRoomJob* Connection::joinRoom(const QString& roomAlias,
     return job;
 }
 
+KnockRoomJob* Connection::knockRoom(const QString& roomAlias,
+                                    const QStringList& serverNames,
+                                    const QString& reason)
+{
+    auto* const job = callApi<KnockRoomJob>(roomAlias, serverNames, reason);
+    // Upon completion, ensure a room object is created in case it hasn't come
+    // with a sync yet. If the room object is not there, provideRoom() will
+    // create it in Join state. finished() is used here instead of success()
+    // to overtake clients that may add their own slots to finished().
+    connect(job, &BaseJob::finished, this, [this, job] {
+        if (job->status().good())
+            provideRoom(job->roomId());
```

Comment:
You really should have read this comment before copying it from `joinRoom()`... `provideRoom()` creates a room in `Join` state by default - this is not the desired behaviour here.
```suggestion
    // with a sync yet. If the room object is not there, provideRoom() will
    // create it in Knock state. finished() is used here instead of success()
    // to overtake clients that may add their own slots to finished().
    connect(job, &BaseJob::finished, this, [this, job] {
        if (job->status().good())
            provideRoom(job->roomId(), JoinState::Knock);
```
There's yet another potential problem here: `provideRoom()` was written with just three join states in mind, and I cannot figure off the top of my head how it will behave if the room is found but is not in the `Leave` state. The code of `provideRoom()` should be reviewed and possibly updated to make sure Quotient clients don't accidentally turn, e.g., already joined rooms into knock rooms (the homeserver won't allow that but the local state may be screwed).